### PR TITLE
F setup logger

### PR DESCRIPTION
**Pull request by ConnorValerio**
Set up winston logging library to replace console.log so that we can utilise logging levels.

Added endpoints to get/set the logging level at runtime.

### CODE REVIEW

File: src/logger.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    winston.format.timestamp({format: "YYYY-MM-DD HH:mm:ss"}),
+    winston.format.printf(info => `[${info.timestamp}] [${info.level.toUpperCase()}]: ${info.message}`)
+);
+
+export const logger = winston.createLogger({
+    format: logFormat,
+    transports: [
+        new winston.transports.File({
+            level: "info",
+            filename: `/var/log/raptor/raptor.log`,
```
Comment (iblocks-Xander) [inline src/logger.ts]: I would say that this log location should be configurable, possibly an environment variable passed through with a default of this location.

Comment (iblocks-Xander) [reply src/logger.ts]: It's also worth noting that this path in unlikely to exist on Windows, so local debugging might be impacted.

File: src/logger.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+const logFormat = winston.format.combine(
+    winston.format.splat(),
+    winston.format.timestamp({format: "YYYY-MM-DD HH:mm:ss"}),
+    winston.format.printf(info => `[${info.timestamp}] [${info.level.toUpperCase()}]: ${info.message}`)
+);
+
+export const logger = winston.createLogger({
+    format: logFormat,
+    transports: [
```
Comment (iblocks-Xander) [inline src/logger.ts:29]: We should add a config option for allowing to log to console as well. Especially for the ability to run and debug locally.

File: src/logger.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        new winston.transports.File({
+            level: "info",
```
Comment (iblocks-Xander) [inline src/logger.ts]: We should also be able to set the starting log level with an environment variable.

File: src/transfer-patterns.ts [hunk 4]
```diff
@@ -31,9 +32,9 @@ export async function run(filename: string) {
 }
 
 if (process.argv[2]) {
-  console.log("I think this runs?");
-  run(process.argv[2]).catch(e => console.error(e));
+  logger.info("I think this runs?");
```
Comment (iblocks-Xander) [inline src/transfer-patterns.ts:35]: We can just get rid of this line!

File: src/runner.ts [hunk 5]
```diff
@@ -1,118 +1,144 @@
...
+
+    app.get("/set-log-level", (req, res, next) => {
+        logger.verbose("Request to '/set-log-level' with req.query: %s", JSON.stringify(req.query));
+
+        const level = req.query.level;
+        if (!level || !logLevels.includes(level)) {
+            res.status(500).send("incorrect/missing query param [level], possible values are: [ " + logLevels + " ]");
+        } else {
+            const prevLevel = logger.transports[0].level;
+            logger.transports[0].level = level;
```
Comment (iblocks-Xander) [inline src/runner.ts]: We should be able to support multiple transports, especially in the scenario of having a console logger as well as file logger.
Maybe a method in the `logger.ts` file would work.

File: src/raptor/RaptorTimeRangeQuery.ts [hunk 6]
```diff
@@ -33,8 +34,7 @@ export class RaptorTimeRangeQuery<T> {
     const midnight = this.getMidnight(startRange);
     const startSeconds = (startRange.valueOf() - midnight.valueOf()) / 1000;
     const endSeconds = (endRange.valueOf() - midnight.valueOf()) / 1000;
-    console.log("Searching from ", startSeconds);
-    console.log("until ", endSeconds);
+    logger.debug("Searching from %o seconds until %o seconds", startSeconds, endSeconds);
```
Comment (iblocks-Xander) [inline src/raptor/RaptorTimeRangeQuery.ts]: The string format should be a bit more precise I think. `startSeconds` and `endSeconds` aren't objects, they are just numbers at this point.

**Review by iblocks-Xander — COMMENTED** [review]
First pass is done.

File: src/logger.ts (hunk 1 above)
Comment (ConnorValerio) [reply src/logger.ts]: 74de274
--

Added. Is it okay to leave the "var/log.." path as the default though?

File: src/logger.ts (hunk 2 above)
Comment (ConnorValerio) [reply src/logger.ts:29]: 74de274
--


Added. Have used the environment variable as a property to silence the Console logger transport if it is not needed.

File: src/logger.ts (hunk 3 above)
Comment (ConnorValerio) [reply src/logger.ts]: 74de274
--


Added.

File: src/transfer-patterns.ts (hunk 4 above)
Comment (ConnorValerio) [reply src/transfer-patterns.ts:35]: If it helps I have never seen this print anyway. So I am unsure if it does run?

File: src/runner.ts (hunk 5 above)
Comment (ConnorValerio) [reply src/runner.ts]: Is it sufficient to set the same log level for all transports, or do you mean being able to individually configure each?

Comment (iblocks-Xander) [reply src/runner.ts]: I think changing the log level for all of the loggers is fine.

File: src/logger.ts (hunk 2 above)
Comment (iblocks-Xander) [reply src/logger.ts:29]: :+1:

File: src/logger.ts (hunk 1 above)
Comment (iblocks-Xander) [reply src/logger.ts]: `/var/log...` is fine to us as a default.

File: src/transfer-patterns.ts (hunk 4 above)
Comment (iblocks-Xander) [reply src/transfer-patterns.ts:35]: I think we don't use it, but we should remove it regardless.

File: src/runner.ts (hunk 5 above)
Comment (ConnorValerio) [reply src/runner.ts]: 1caac87
--


 Added getters/setters to change the log level for all Logger Transports.

File: src/raptor/RaptorTimeRangeQuery.ts (hunk 6 above)
Comment (ConnorValerio) [reply src/raptor/RaptorTimeRangeQuery.ts]: 8c6b938
--


 Changed type in log message string from object to number.

File: src/logger.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
+const winston = require("winston");
+
+// list of logging levels in order
+export const logLevels = ["error", "warn", "info", "verbose", "debug", "silly"];
+
+// read in environment variables
+let logLevel = process.env.RAPTOR_LOG_LEVEL && logLevels.includes(process.env.RAPTOR_LOG_LEVEL) ? process.env.RAPTOR_LOG_LEVEL : "info";
+const logFile = process.env.RAPTOR_LOG_LOCATION || "/var/log/raptor/raptor.log";
+const logToConsole = process.env.RAPTOR_LOG_TO_CONSOLE || false;
```
Comment (iblocks-Xander) [inline src/logger.ts:9]: Having just run into this, can we turn the console log on by default? The failure was completely silent as the console output nothing!

Comment (ConnorValerio) [reply src/logger.ts:9]: 209959d
--
Added.

**Pull request merged**